File: cpp/quick_sort.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <chrono>

struct Data {
    int number;
    std::string letters;

    bool operator<=(const Data& other) const {
        return number <= other.number;
    }
};

void quickSort(std::vector<Data>& arr, int low, int high);
int partition(std::vector<Data>& arr, int low, int high);

void quickSort(std::vector<Data>& arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int partition(std::vector<Data>& arr, int low, int high) {
    Data pivot = arr[high];
    int i = low - 1;
    for (int j = low; j < high; ++j) {
        if (arr[j] <= pivot) {
            ++i;
            std::swap(arr[i], arr[j]);
        }
    }
    std::swap(arr[i + 1], arr[high]);
    return i + 1;
}

int main() {
    std::string filePath;
    std::cout << "Enter the CSV file path: ";
    std::getline(std::cin, filePath);

    std::vector<Data> dataList;
    std::ifstream infile(filePath);
    if (!infile) {
        std::cout << "Error reading file: cannot open file\n";
        return 1;
    }
    std::string line;
    while (std::getline(infile, line)) {
        std::istringstream ss(line);
        std::string numberStr, letters;
        if (std::getline(ss, numberStr, ',') && std::getline(ss, letters)) {
            try {
                int number = std::stoi(numberStr);
                dataList.push_back({number, letters});
            } catch (...) {
                std::cout << "Error reading file: invalid number format\n";
                return 1;
            }
        }
    }
    infile.close();

    auto startTime = std::chrono::high_resolution_clock::now();
    quickSort(dataList, 0, static_cast<int>(dataList.size()) - 1);
    auto endTime = std::chrono::high_resolution_clock::now();
    double durationMs = std::chrono::duration<double, std::milli>(endTime - startTime).count();
    std::cout << "Sorting runtime: " << durationMs << " ms\n";

    std::string outputFileName = "quick_sort_" + std::to_string(dataList.size()) + ".csv";
    std::ofstream outfile(outputFileName);
    if (!outfile) {
        std::cout << "Error writing file: cannot open file\n";
        return 1;
    }
    for (const auto& d : dataList) {
        outfile << d.number << "," << d.letters << "\n";
    }
    outfile.close();

    return 0;
}

File: java/quickSort.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;

public class quickSort {
    // Data class defined within the same file
    static class Data implements Comparable<Data> {
        int number;
        String letters;

        public Data(int number, String letters) {
            this.number = number;
            this.letters = letters;
        }

        @Override
        public int compareTo(Data other) {
            return Integer.compare(this.number, other.number);
        }
    }

    public static void quickSort(Data[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    private static int partition(Data[] arr, int low, int high) {
        Data pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr[j].compareTo(pivot) <= 0) {
                i++;
                Data temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        Data temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        return i + 1;
    }

    public static void main(String[] args) {
        java.util.Scanner scanner = new java.util.Scanner(System.in);
        System.out.print("Enter the CSV file path: ");
        String filePath = scanner.nextLine();
        ArrayList<Data> dataList = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",", 2);
                if (parts.length == 2) {
                    int number = Integer.parseInt(parts[0].trim());
                    String letters = parts[1].trim();
                    dataList.add(new Data(number, letters));
                }
            }
        } catch (IOException | NumberFormatException e) {
            System.out.println("Error reading file: " + e.getMessage());
            return;
        }
        Data[] arr = dataList.toArray(new Data[dataList.size()]);

        long startTime = System.nanoTime();
        quickSort(arr, 0, arr.length - 1);
        long endTime = System.nanoTime();
        double durationMs = (endTime - startTime) / 1_000_000.0;
        System.out.printf("Sorting runtime: %.3f ms%n", durationMs);

        String outputFileName = "quick_sort_" + arr.length + ".csv";
        try (java.io.BufferedWriter bw = new java.io.BufferedWriter(new java.io.FileWriter(outputFileName))) {
            for (Data d : arr) {
                bw.write(d.number + "," + d.letters);
                bw.newLine();
            }
        } catch (IOException e) {
            System.out.println("Error writing file: " + e.getMessage());
        }
    }
}
